File: solutions/free-pascal/proverb/1/Proverb.pas
unit Proverb;

{$mode ObjFPC}{$H+}

interface

function recite(const strings : Array Of String) : string;

implementation

function recite(const strings : Array Of String) : string;
var
  i : integer;
begin
  result := '';
  if length(strings) = 0 then
    exit();

  for i := 1 to high(strings) do
    result := result + 'For want of a ' + strings[i - 1] + ' the ' + strings[i] + ' was lost.' + #10;

  result := result + 'And all for the want of a ' + strings[0] + '.';
end;

end.

File: solutions/swift/proverb/1/Sources/Proverb/Proverb.swift
struct Proverb {
    let verses: [String]

    init(_ wants: [String]) {
        verses = Self.buildVerses(for: wants)
    }

    func recite() -> String {
        verses.joined(separator: "\n")
    }

    private static func buildVerses(for wants: [String]) -> [String] {
        guard let first = wants.first else {
            return []
        }

        let mainVerses = zip(wants, wants.dropFirst()).map { (current, next) in
            "For want of a \(current) the \(next) was lost."
        }

        return mainVerses + ["And all for the want of a \(first)."]
    }
}